Add a mode for erasing erasable jane-syntax

antalsz opened the pull request:
We will use this mode to prepare our public-release code.  We still get a weaker round-tripping check -- we check that the resulting trees are equal _except_ for `jane.erasable.*` attributes.

antalsz committed:
Add (untested) support for erasing erasable Jane syntax

Signed-off-by: Antal Spector-Zabusky <antal.b.sz@gmail.com>

antalsz committed:
Make normalization ignore erasable jane syntax if requested

Signed-off-by: Antal Spector-Zabusky <antal.b.sz@gmail.com>

antalsz committed:
Add failing test of erasure

Signed-off-by: Antal Spector-Zabusky <antal.b.sz@gmail.com>

antalsz committed:
Update parser-standard to antalsz/flambda-backend 5b23940cfccd3047

Signed-off-by: Antal Spector-Zabusky <antal.b.sz@gmail.com>

antalsz committed:
Update parser-standard and location to antalsz/flambda-backend c056f5f2b0082711b596337f270a8eccf06d0b06

Signed-off-by: Antal Spector-Zabusky <antal.b.sz@gmail.com>

goldfirere commented on lib/bin_conf/Bin_conf.ml:
```diff
@@ -325,6 +327,15 @@ let ocp_indent_config =
 
+let erase_jane_syntax =
+  let doc =
+    "Erase all erasable Jane Street syntax extensions.  THIS OPTION WILL \
```
I would include something about public release in the docstring.

goldfirere commented on lib/erase_jane_syntax/erase_jane_syntax.ml:
```diff
@@ -0,0 +1,5 @@
+let should_erase_ref = ref false
```
If this stuff were right in `Bin_conf`, you could just access `global_conf` directly, right? Seems simpler -- and this very short module (and directory!) seems heavier than necessary.

Unless there are e.g. merge reasons to separate?

goldfirere commented on lib/erase_jane_syntax/erase_jane_syntax.mli:
```diff
@@ -0,0 +1,3 @@
+val set_should_erase : bool -> unit
```
Do please comment these.

goldfirere commented on vendor/parser-extended/dune:
```diff
@@ -3,7 +3,12 @@
   (:standard -w -9 -open Parser_shims -open Ocaml_common))
- (libraries compiler-libs.common menhirLib parser_shims ocaml_common))
+ (libraries
+  erase_jane_syntax
```
Ah -- is this the reason it all has to be in its own module? That makes some sense. Would it be better just to pass the bool to the exported functions in `Parser_extended.Parse`? That wouldn't be hard -- it could be handled by the `wrap` function. But perhaps no better than the status quo here.

In any case, please briefly document the need for the separate library.

antalsz replied:
Yes, that's right -- I didn't think to look closely at passing it in like that, since I was concerned about Menhir entry points, but I'll take a look, possibly make the change, and document things either way.

antalsz replied:
I'm inclined to leave it as is: I need to set something up that's available to Menhir and then set globally, and this same thing also has to be available in `Ast_helper` or I have to propagate booleans there, and given all that (plus the merge benefits), there doesn't seem to be much of a real locality win?

goldfirere commented on Makefile:
```diff
@@ -33,6 +33,7 @@ regtests:
 
 regtests-promote:
 	dune runtest --auto-promote
+	dune build @help --auto-promote
```
This looks like part of that other PR. Did you mean to put it here? What does this do?

antalsz replied:
It is, just means I need to rebas

antalsz replied:
This is from the draft PR #33, I figured I would rebase once that was merged -- this promotes the `ocamlformat-help.txt` and `ocamlformat-rpc-help.txt` files, which ought to be kept in sync with what `ocamlformat` actually outputs; indeed, the CI checks this.  However, I'm having an issue where I have a different version of `cmdliner` locally that's changing the output, hence #33, but I didn't get that working.

antalsz replied:
Update: Now #33 is simpler, #34 has this, and I've taken this line out of here.

goldfirere commented on vendor/parser-standard/jane_syntax_parsing.mli:
```diff
@@ -98,7 +98,7 @@ module Feature : sig
 
   (** The component of an attribute or extension name that identifies the
-      feature. This is third component.
+      feature. This is the third component.
```
I assume these parser-standard changes are just copied over from your branch -- let me know if that's not the case.

antalsz replied:
That's correct, those files are literally `cp`ed over.

goldfirere reviewed (commented):
As we discussed, it would be good to change the safety check to remove `jane.erasable` only from the original source. The reformatted source should already have all of these removed. So changing the check not to remove attributes from the reformatted source will ensure that we haven't forgotten anything here.

antalsz committed:
Update `--erase-jane-syntax` documentation per review

Signed-off-by: Antal Spector-Zabusky <antal.b.sz@gmail.com>

antalsz committed:
Documentation for `Erase_jane_syntax`

Signed-off-by: Antal Spector-Zabusky <antal.b.sz@gmail.com>

antalsz committed:
Only remove erasable `Jane_syntax` attributes from the old AST

Signed-off-by: Antal Spector-Zabusky <antal.b.sz@gmail.com>

antalsz commented:
I think I've now addressed all of your questions except for the #33 thing

antalsz commented:
@goldfirere I think I've addressed all of your concerns; I went with keeping `Erase_jane_syntax` as it is, but if you feel strongly, I could be argued in the other direction.  I think it's close.

goldfirere commented:
Looks pretty good to me -- just some CI to clean up. Thanks!

antalsz committed:
Fix formatting

Signed-off-by: Antal Spector-Zabusky <antal.b.sz@gmail.com>
